networkx/algorithms/distance_measures.py

```diff
@@ -769,3 +770,104 @@ def resistance_distance(G, nodeA, nodeB, weight=None, invert_weight=True):
     rd = Ldet * LdiagAB_s / LdiagA_s
 
     return rd
+
+
+@nx._dispatch(edge_attrs="weight")
+def kemeny_constant(G, weight=None):
+    """Returns Kemeny's constant of the given graph.
+
+    *Kemeny's constant* of a graph `G` can be computed by regarding the graph
+    as a Markov chain. Kemeny's constant is then the expected number of time steps
+    to transition from a starting state i to a random destination state
+    sampled from the Markov chain's stationary distribution.
+    Kemeny's constant is independent of the chosen initial state [1]_.
+
+    Kemeny's constant is used as a measure of connectivity of graphs.
+    Large values indicate a poorly connected graph whereas low values
+    correspond to well-connected graphs.
+
+    If weight is not provided, then a weight of 1 is used for all edges.
+
+    Parameters
+    ----------
+    G : NetworkX graph
+
+    weight : string or None, optional (default=None)
+       The edge data key used to compute Kemeny's constant.
+       If None, then each edge has weight 1.
+
+    Returns
+    -------
+    K : float
+        Kemeny's constant of the graph `G`.
+
+    Raises
+    ------
+    NetworkXError
+        If the graph `G` is directed.
+
+    NetworkXError
+        If the graph `G` is not connected.
+
+    NetworkXError
+        If the graph `G` does not contain any nodes.
+
+    NetworkXError
+        If the graph `G` contains edges with negative weights.
+
+    Examples
+    --------
+    >>> G = nx.complete_graph(5)
+    >>> round(nx.kemeny_constant(G), 10)
+    3.2
+
+    Notes
+    -----
+    The implementation is based on equation (3.3) in [2]_. Self-loops are included.
+    Multi-edges are contracted in one edge with weight equal to the sum of the weights.
+
+    References
+    ----------
+    .. [1] Wikipedia
+       "Kemeny's constant."
+       https://en.wikipedia.org/wiki/Kemeny%27s_constant
+    .. [2] Lovász L.
+        Random walks on graphs: A survey.
+        Paul Erdös is Eighty, vol. 2, Bolyai Society,
+        Mathematical Studies, Keszthely, Hungary (1993), pp. 1-46
+    """
+    import numpy as np
+    import scipy as sp
+
+    if len(G) == 0:
+        msg = "Graph G must contain at least one node."
+        raise nx.NetworkXError(msg)
+    elif nx.is_directed(G):
```

Comment:
These `elif` should just be `if` (because the previous `if` will always `raise` and thus stop the program flow.

Also, many of these error messages should be included in the raise call (removing the variable `msg`, but more importantly putting it all on one line -- that gets shown by the error traceback)

Reply:
First part, done.

Second part, I am not completely aware of your question, but I presume it amounts to changing
`msg = "Graph G must contain at least one node.";
raise nx.NetworkXError(msg)`

to

`raise nx.NetworkXError("Graph G must contain at least one node.")`

which I did